### PR TITLE
Implement OnNegotiationNeeded

### LINKED ISSUE #817: Implement OnNegotiationNeeded
This should be added to support the Tor folks!

### PR DESCRIPTION
**Pull request by Sean-Der**
This event is fired when a change has occurred
which requires session negotiation.

Resolves #817

### CODE REVIEW

**Comment by jech** [comment]
The change to operations.start is not correct.  The original code was carefully written to continue the operations queue if an operation panics; the new code will stop the operations queue if an operation panics.  I'm afraid that stopping the queue will lead to difficult to debug deadlocks.

The `ops.done` mechanism feels very ad-hoc to me, and I'm not sure whether it's needed.  Instead of special-casing `done`:
```
ops.done = f
```
couldn't we simply enqueue an operation that recursively reenqueues itself as long as the queue is not empty?
```
g := func() {
    if !ops.isEmpty() {
        ops.Enqueue(g)
        return
    }
    f()
}
ops.Enqueue(g)
```

Finally, I'm not sure I understand how the code ensures that `OnNegotiationNeeded` is eventually triggered.  If a client keeps enqueuing new ops, then the queue will never drain.  Are there guarantees that this kind of livelock never happens?

**Comment by tekig** [comment]
left to do in pull #1322

1. Maybe, the renegotiation does not understand when the stream is closed. Will always call him.
pion, does not close the connection, but makes it inactive? need to check

2. I'm sure [this code](https://github.com/pion/webrtc/pull/1322#issuecomment-663878012) can be written better

3. Chrome closes the transceivers as the spec says, but pion makes it inactive. Or you need to remove the [check for closed](https://github.com/pion/webrtc/pull/1322/commits/a06e8a27e06e3bfb24716636a3f5d868935ede7e#diff-d2ee665bb8c62c914fee5e9c743be25cR332-R334) transceivers in peon or implement transceivers closure, as in the specification. 

4. More tests)

It might have been faster to write code than to write this comment

**Comment by Sean-Der** [comment]
@tekig i added you to the Pion GitHub organization so you can push to the branch directly! Sorry about cherry-picking I though I would have more time

**Comment by Sean-Der** [comment]
https://github.com/pion/webrtc/commit/4aa50a36be648181f5512e0b551141a0e166b30f I did take this and squashed.

**Comment by Sean-Der** [comment]
@tekig please take w/e you can and I can help tomorrow. This is my #1 priority for the next few days!

**Comment by tekig** [comment]
With an example, it will be easier to test with a browser. Didn't use websocket to not include dependencies.
I hope for you, I will help if possible

**Comment by tekig** [comment]
#1361

**Comment by tekig** [comment]
Done, test with an example. Especially in the process of deleting a track, it differs from the specification and some steps are skipped due to the peculiarities of the pion code
Sometimes an error appears in the browser, but it is not related to OnNegotiationNeeded
Chrome
`DOMException: Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to set remote answer sdp: Failed to apply the description for 0: Failed to set SSL role for the transport.`
Firefox
`DOMException: Answer tried to set recv when offer did not set sen`

**Comment by Sean-Der** [comment]
> `DOMException: Answer tried to set recv when offer did not set send`

We should fix this now. It is really easy actually! https://github.com/pion/webrtc/issues/717 and https://github.com/pion/webrtc/pull/1199 

I will take a look at that tomorrow.

> `Failed to set SSL role for the transport`

My guess is we are doing the wrong value for `setup:` during renegotation. I don't know the answer around this. It will just be a signaling/text change though!

File: peerconnection.go [hunk 1]
```diff
@@ -242,6 +245,154 @@ func (pc *PeerConnection) OnDataChannel(f func(*DataChannel)) {
...
+
+	if localDesc == nil {
+		return true
+	}
+
+	if pc.sctpTransport == nil {
+		return true
+	}
+
+	if len(pc.sctpTransport.dataChannels) != 0 {
```
Comment (Sean-Der) [inline peerconnection.go:335]: Is this just asserting that you have requested a DataChannel, but the Transport hasn't been started?

Can you use the `SCTPTransport.State()` instead?

Comment (Sean-Der) [reply peerconnection.go:335]: Never mind I see this is clearly defined in the RFC.

Can you make this a helper function in `sdp.go`? Then we can re-use it. We scan a SDP for `Does it have a datachannel a few places`

**Comment by Sean-Der** [comment]
@tekig this looks really great to me (for merging into master) Last few things.

* Can you fix the failures in CI? 
    - `go test -race`
    - `golangci-lint`
* Can you do a rebase and squash everything into one commit?

@jech @jeremija you see anything that really stands out? I think after we land I think we can simplify the examples a bit. They are really great, just a bit verbose. Also maybe revisit `negotiationNeededState`.

Really great work on this again. This is a huge undertaking for first contribution.

**Comment by tekig** [comment]
> Can you fix the failures in CI?
> * `go test -race`
> * `golangci-lint`

`golangci-lint`  [reports](https://travis-ci.org/github/pion/webrtc/jobs/717978296#L356-L364) the error in only one place. I decided to make the method more general, so that it will be used more often, if necessary.

**Pull request merged**